src/embind/embind.js

```diff
@@ -1734,38 +1734,54 @@ var LibraryEmbind = {
     }
   },
 
-  $finalizationGroup: false,
+  $finalizationRegistry: false,
 
-  $detachFinalizer_deps: ['$finalizationGroup'],
+  $detachFinalizer_deps: ['$finalizationRegistry'],
   $detachFinalizer: function(handle) {},
 
-  $attachFinalizer__deps: ['$finalizationGroup', '$detachFinalizer',
+  $attachFinalizer__deps: ['$finalizationRegistry', '$detachFinalizer',
                            '$releaseClassHandle'],
   $attachFinalizer: function(handle) {
-    if ('undefined' === typeof FinalizationGroup) {
-        attachFinalizer = (handle) => handle;
+    if ('undefined' === typeof FinalizationRegistry) {
+        attachFinalizer = (handle) => { return handle; };
         return handle;
     }
-    // If the running environment has a FinalizationGroup (see
+    // If the running environment has a FinalizationRegistry (see
     // https://github.com/tc39/proposal-weakrefs), then attach finalizers
-    // for class handles.  We check for the presence of FinalizationGroup
+    // for class handles.  We check for the presence of FinalizationRegistry
     // at run-time, not build-time.
-    finalizationGroup = new FinalizationGroup(function (iter) {
-        for (var result = iter.next(); !result.done; result = iter.next()) {
-            var $$ = result.value;
-            if (!$$.ptr) {
-                console.warn('object already deleted: ' + $$.ptr);
-            } else {
-                releaseClassHandle($$);
-            }
-        }
+    finalizationRegistry = new FinalizationRegistry(function (info) {
+#if ASSERTIONS
+        console.warn(info.leakWarning.stack.replace(/^Error: /, ''));
+#endif
+        releaseClassHandle(info.$$);
     });
     attachFinalizer = (handle) => {
-        finalizationGroup.register(handle, handle.$$, handle.$$);
+        var $$ = handle.$$;
+        var hasSmartPtr = !!$$.smartPtr;
+        if (hasSmartPtr) {
+            // We should not call the destructor on raw pointers in case other code expects the pointee to live
+            var info = { $$: $$ };
+#if ASSERTIONS
+            // Create a warning as an Error instance in advance so that we can store
+            // the current stacktrace and point to it when / if a leak is detected.
+            // This is more useful than the empty stacktrace of `FinalizationRegistry`
+            // callback.
+            var cls = $$.ptrType.registeredClass;
+            info.leakWarning = new Error("Embind found a leaked C++ instance " + cls.name + " <0x" + $$.ptr.toString(16) + ">.\n" +
+            "We'll free it automatically in this case, but this functionality is not reliable across various environments.\n" +
+            "Make sure to invoke .delete() manually once you're done with the instance instead.\n" +
+            "Originally allocated"); // `.stack` will add "at ..." after this sentence
+            if ('captureStackTrace' in Error) {
+                Error.captureStackTrace(info.leakWarning, cls.constructor);
```

Comment:
Btw I now remember there were some cases (some Embind APIs) where `cls.constructor` didn't exist, and we need to choose some different baseline. @mmarczell-graphisoft could you look into this maybe? Essentially try 1) creating an instance from C++ side and returning from a function exposed via Embind 2) creating an instance from JS side from a class exposed via Embind.

In both cases stacktrace should contain where object was created.

Reply:
@RReverser That seems quite difficult to do generally. In my testing the problem is not that `cls.constructor` doesn't exist, but that it is not a part of the stack trace. Since the meaning of this parameter is roughly "omit stack frames above this" (see [v8.dev blog post](https://v8.dev/docs/stack-trace-api#stack-trace-collection-for-custom-exceptions)), if the function is not in the stack trace, the result is an empty stack trace. Here's two stack traces for illustration, the first is for an object being returned from a function and the second is an object being constructed from JS. Both were captured with this second parameter omitted, so nothing was filtered out.

```
main.js:1693 Embind found a leaked C++ instance Bar <0x505ec4>.
We'll free it automatically in this case, but this functionality is not reliable across various environments.
Make sure to invoke .delete() manually once you're done with the instance instead.
Originally allocated
    at attachFinalizer (http://localhost:8000/main.js:1706:11)
    at makeClassHandle (http://localhost:8000/main.js:2049:9)
    at RegisteredPointer_fromWireType [as fromWireType] (http://localhost:8000/main.js:2105:10)
    at Object.bar (eval at new_ (http://localhost:8000/main.js:2499:22), <anonymous>:11:19)
    at Object.Module.onRuntimeInitialized (http://localhost:8000/index.html:12:25)
    at doRun (http://localhost:8000/main.js:6845:69)
    at run (http://localhost:8000/main.js:6858:3)
    at runCaller (http://localhost:8000/main.js:6820:18)
    at removeRunDependency (http://localhost:8000/main.js:1107:4)
    at receiveInstance (http://localhost:8000/main.js:1221:3)

main.js:1693 Embind found a leaked C++ instance Foo <0x505e34>.
We'll free it automatically in this case, but this functionality is not reliable across various environments.
Make sure to invoke .delete() manually once you're done with the instance instead.
Originally allocated
    at attachFinalizer (http://localhost:8000/main.js:1706:11)
    at attachFinalizer (http://localhost:8000/main.js:1715:9)
    at makeClassHandle (http://localhost:8000/main.js:2049:9)
    at RegisteredPointer_fromWireType [as fromWireType] (http://localhost:8000/main.js:2105:10)
    at Foo.constructor$Foo (eval at new_ (http://localhost:8000/main.js:2499:22), <anonymous>:8:19)
    at Foo.<anonymous> (http://localhost:8000/main.js:2276:16)
    at new Foo (eval at createNamedFunction (http://localhost:8000/main.js:1495:9), <anonymous>:4:34)
    at Object.Module.onRuntimeInitialized (http://localhost:8000/index.html:10:18)
    at doRun (http://localhost:8000/main.js:6845:69)
    at run (http://localhost:8000/main.js:6858:3)
```
To guarantee a clean non-empty stack trace, my only idea is create the stack trace like the above, without providing this second parameter, and then string manipulating it into the desired form. But that seems like overkill. I see no other way of sensibly figuring out "where from" we are being called.

Reply:
> In my testing the problem is not that `cls.constructor` doesn't exist, but that it is not a part of the stack trace.

Oh right, _that_ was the problem (sorry, I forgot the details).

I agree that would be an overkill. I suggest to instead just take a function a bit higher up the stack - I believe the `RegisteredPointer_fromWireType` one should be always present for this purpose. You just need to check how to retrieve it, as it's likely stored as a property of some descriptor.

Reply:
That seems better, at least stacktrace is present in both cases and the top of the stacktrace still clearly points to the culprit.